### PR TITLE
Die with dignity

### PR DESCRIPTION
**Pull request by jasontedor**
Today when a thread encounters a fatal unrecoverable error that
threatens the stability of the JVM, Elasticsearch marches on. This
includes out of memory errors, stack overflow errors and other errors
that leave the JVM in a questionable state. Instead, the Elasticsearch
JVM should die when these errors are encountered. This commit causes
this to be the case.

Relates #19231

### CODE REVIEW

**Comment by rmuir** [comment]
+1

**Comment by tlrx** [comment]
LGTM, I like it

**Comment by tlrx** [comment]
We might want to document the exit codes somewhere too

**Comment by jasontedor** [comment]
> We might want to document the exit codes somewhere too

@tlrx I think that's a good idea, but I'm unsure of a good place to document it. Do you or @clintongormley have any ideas?

**Comment by dadoonet** [comment]
FYI we documented something similar for plugins: https://www.elastic.co/guide/en/elasticsearch/plugins/current/_other_command_line_parameters.html

**Comment by clintongormley** [comment]
@jasontedor not perfect, but perhaps a section under [Setup](https://www.elastic.co/guide/en/elasticsearch/reference/master/setup.html)?

File: docs/reference/setup/stopping.asciidoc [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from the cluster, sync translogs to disk, and perform other related cleanup activities. You can help
+ensure an orderly shutdown by properly stopping Elasticsearch.
+
+If you're running Elasticsearch as a service, you can stop Elasticsearch via the service management
+functionality provided by your installation.
+
+If you're running Elasticsearch directly, you can stop Elasticsearch by sending control-C if you're
+running Elasticsearch in the console, or by sending `SIGTERM` to the Elasticsearch process on a
+POSIX system. You can obtain the PID to send the signal to via various tools (e.g., `ps` or `jps`)
+or by specifying a location to write a PID file to on startup (`-p <path>`).
```
Comment (tlrx) [inline docs/reference/setup/stopping.asciidoc]: When ran directly the pid is shown in the logs/stdout too.

Comment (jasontedor) [inline docs/reference/setup/stopping.asciidoc]: Thanks @tlrx, I pushed 94ef28b1eb3afadd669b4abc828acb34f59d4d56.

**Comment by jasontedor** [comment]
@tlrx @clintongormley I pushed docs in 74c1708c9e3aac2badf0f8ddb886235d19fc8321; could you take a look?

File: docs/reference/setup/stopping.asciidoc (hunk 1 above)
Comment (tlrx) [inline docs/reference/setup/stopping.asciidoc]: Sorry, I should be more clear. I think we can rephrase like:

> Elasticsearch indicates the PID in logs or standard ouput during startup:
> [INFO ][node                     ] [Araña] version[2.3.3], pid[6674], build[218bdf1/2016-05-17T15:40:04Z]
> You can also obtain...

or something like this

**Comment by tlrx** [comment]
LGTM

File: docs/reference/setup/stopping.asciidoc (hunk 1 above)
Comment (jasontedor) [inline docs/reference/setup/stopping.asciidoc]: I pushed 45501405f25b461f9b93e3ff5c4f45faa62ac4cf. Is that more like what you had in mind?

Comment (tlrx) [inline docs/reference/setup/stopping.asciidoc]: https://github.com/elastic/elasticsearch/commit/45501405f25b461f9b93e3ff5c4f45faa62ac4cf is good, thanks!

**Comment by clintongormley** [comment]
Sorry I'm late to this.  You can also get the PID from the node info API:

    curl -XGET "http://192.168.1.10:9200/_nodes/process"

    {
      "_nodes": {
        "total": 1,
        "successful": 1,
        "failed": 0
      },
      "cluster_name": "elasticsearch",
      "nodes": {
        "uOPHTfzUQUyGRuxdSUBniQ": {
          "name": "Karl Lykos",
          "transport_address": "127.0.0.1:9300",
          "host": "127.0.0.1",
          "ip": "127.0.0.1",
          "version": "5.0.0-alpha4",
          "build_hash": "3f5b994",
          "http_address": "127.0.0.1:9200",
          "roles": [
            "master",
            "data",
            "ingest"
          ],
          "process": {
            "refresh_interval_in_millis": 1000,
            "id": 26283,  ##### HERE #####
            "mlockall": false
          }
        }
      }
    }

**Comment by jasontedor** [comment]
@clintongormley Yes, and the cat nodes API too. The concern I have there is that that will list all the nodes in the cluster when only the local node is needed and that leads to more effort on the end-user? I'm happy to add these though if you think otherwise. 😄

**Pull request merged**